pkg/database/dbaas.go

```diff
@@ -0,0 +1,120 @@
+)
+
+// DBaaSDatabase contains the Database struct needed with the plan, specific of DBaaS
+type DBaaSDatabase struct {
+	Database
+	plan *string
+}
+
+// Execute starts the saving process of the data in the billing database
+func (d *DBaaSDatabase) Execute(dctx *Context) error {
+	p := pipeline.NewPipeline[*Context]()
+	p.WithSteps(
+		p.NewStep("Open database connection", d.openConnection),
+		p.WithNestedSteps("Save initial billing configuration", nil,
+			p.NewStep("Begin transaction", d.beginTransaction),
+			p.NewStep("Ensure initial billing database configuration", d.ensureInitConfiguration),
+			p.NewStep("Commit transaction", d.commitTransaction),
+		).WithErrorHandler(d.rollback),
+		p.NewStep("Save DBaaS usage to billing database", d.saveUsageToDatabase),
+		p.NewStep("Close database connection", d.closeConnection),
+	)
+	return p.RunWithContext(dctx)
+}
+
+// saveUsageToDatabase saves each previously aggregated DBaaS to the billing database
+func (d *DBaaSDatabase) saveUsageToDatabase(dctx *Context) error {
+	log := ctrl.LoggerFrom(dctx)
+	for _, aggregated := range *dctx.AggregatedObjects {
+		err := d.ensureAggregatedServiceUsage(dctx, aggregated)
+		if err != nil {
+			log.Error(err, "Cannot save aggregated DBaaS service record to billing database")
+			continue
+		}
+	}
+	return nil
+}
+
+// ensureAggregatedServiceUsage saves the aggregated database service usage by namespace and plan to the billing database
+// To save the correct data to the database the function also matches a relevant product, discount (if any) and query.
+func (d *DBaaSDatabase) ensureAggregatedServiceUsage(dctx *Context, aggregatedDatabaseService Aggregated) error {
+	log := ctrl.LoggerFrom(dctx)
+	tokens, err := aggregatedDatabaseService.DecodeKey()
+	if err != nil {
+		return fmt.Errorf("cannot decode key namespace-plan-dbtype - %s, organization %s, number of instances %f: %w",
+			aggregatedDatabaseService.Key,
+			aggregatedDatabaseService.Organization,
+			aggregatedDatabaseService.Value,
+			err)
+	}
+	namespace := tokens[0]
+	plan := tokens[1]
+	dbType := tokens[2]
+
+	log.Info("Saving DBaaS usage", "namespace", namespace, "plan", plan, "type", dbType, "quantity", aggregatedDatabaseService.Value)
+
+	dctx.organization = &aggregatedDatabaseService.Organization
+	dctx.namespace = &namespace
+	dctx.value = &aggregatedDatabaseService.Value
+	dctx.Aggregated = &aggregatedDatabaseService
+	d.plan = &plan
+
+	d.sourceString = dbaasSourceString{
+		Query:        billingTypes[dbType],
+		Organization: *dctx.organization,
+		Namespace:    namespace,
+		Plan:         plan,
+	}
+
+	p := pipeline.NewPipeline[*Context]()
+	p.WithSteps(
+		p.WithNestedSteps(fmt.Sprintf("Saving DBaaS usage namespace %s", namespace), nil,
+			p.NewStep("Begin database transaction", d.beginTransaction),
+			p.NewStep("Ensure necessary models", d.ensureModels),
+			p.NewStep("Get best match", d.getBestMatch),
+			p.WithNestedSteps("Save to billing database", isFactsUpdatable,
```

Comment:
Just to be sure, the nested step is only executed if the first one returns true?

Reply:
yes